packages/zoe/src/state.js

```diff
@@ -105,46 +105,45 @@ const makeOfferTable = () => {
 const makePayoutMap = () => makeStore('offerHandle');
 
 // Issuer Table
-// Columns: issuer | brand | purse | amountMath
+// Columns: brand | issuer | purse | amountMath
+//
+// The IssuerTable is keyed by brand, but the Issuer is required in order for
+// getPromiseForIssuerRecord() to initialize the records. When
+// getPromiseForIssuerRecord is called and the record doesn't exist, it stores a
+// promise for the record in issuersInProgress, and then builds the record. It
+// updates the tables when done.
 const makeIssuerTable = () => {
   // TODO: make sure this validate function protects against malicious
-  // misshapen objects rather than just a general check.
+  //  misshapen objects rather than just a general check.
   const validateSomewhat = makeValidateProperties(
-    harden(['issuer', 'brand', 'purse', 'amountMath']),
+    harden(['brand', 'issuer', 'purse', 'amountMath']),
   );
 
-  /**
-   * Brand to Issuer Map
-   * @returns {import('@agoric/store').Store<Brand,
-   * Issuer>} Store
-   */
-  const brandToIssuerMap = makeStore('brand');
-
   const makeCustomMethods = table => {
     const issuersInProgress = makeStore();
+    const issuerToBrand = makeStore();
 
     const customMethods = harden({
-      getPurseKeywordRecord: issuerKeywordRecord => {
-        const purseKeywordRecord = {};
-        Object.keys(issuerKeywordRecord).forEach(keyword => {
-          purseKeywordRecord[keyword] = table.get(
-            issuerKeywordRecord[keyword],
-          ).purse;
-        });
-        return harden(purseKeywordRecord);
-      },
-
-      // `issuerP` may be a promise, presence, or local object
+      // `issuerP` may be a promise, presence, or local object. We're going
+      // to get the brand first, since that may be all we need. If there's
+      // already a record, or already a promise for a record, return (a promise
+      // for) that. Otherwise wrap a promise around building the record.
       getPromiseForIssuerRecord: issuerP => {
         return Promise.resolve(issuerP).then(issuer => {
-          if (!table.has(issuer)) {
-            if (issuersInProgress.has(issuer)) {
-              // a promise which resolves to the issuer record
-              return issuersInProgress.get(issuer);
-            }
+          if (issuerToBrand.has(issuer)) {
```

Comment:
The reason I mention the pattern here is because it *is* directly applicable here. Here are the changes that I had previously suggested:
<img width="533" alt="Screen Shot 2020-06-05 at 6 34 17 PM" src="https://user-images.githubusercontent.com/2441069/83933182-3918a380-a75b-11ea-830f-c0a63866b045.png">

Reply:
Yep. I generated that version and compared it to the version above.  The version above seems much more straightforward. 

if there's an entry in the main table
   return it
else if there's an entry in the inProgress table
  return it
else
  build the entries.

vs

if there's nothing in the main table
  if there's nothing in the inProgress table
     build the entries
  return the entry from the inProgress table
return the entry from the main table

The second is far less local, and the consequences of the tests are harder to connect to the tests.